les/utils/limiter.go

```diff
@@ -0,0 +1,395 @@
+		reqWeight = 1
+	}
+	if reqWeight > l.maxWeight {
+		l.maxWeight = reqWeight
+	}
+
+	if nq, ok := l.nodes[id]; ok {
+		nq.queue = append(nq.queue, request{process, reqWeight})
+		nq.sumWeight += reqWeight
+		nq.value = value
+		if address != nq.address {
+			// known id sending request from a new address, move to different address group
+			l.removeFromGroup(nq)
+			l.addToGroup(nq, address)
+		}
+	} else {
+		nq := &nodeQueue{
+			queue:      []request{{process, reqWeight}},
+			id:         id,
+			value:      value,
+			nextWeight: reqWeight,
+			sumWeight:  reqWeight,
+			groupIndex: -1,
+		}
+		nq.flatWeight, nq.valueWeight = selectionWeights(float64(reqWeight)/float64(l.maxWeight), value)
+		if len(l.nodes) == 0 {
+			l.cond.Signal()
+		}
+		l.nodes[id] = nq
+		if nq.valueWeight != 0 {
+			l.valueSelect.Update(nq)
+		}
+		l.addToGroup(nq, address)
+	}
+	l.sumWeight += reqWeight
+	if l.sumWeight > l.sumWeightLimit {
+		l.dropRequests()
+	}
+	return process
+}
+
+// update updates the selection weights of the node queue
+func (l *Limiter) update(nq *nodeQueue) {
+	flatWeight, valueWeight := selectionWeights(float64(nq.nextWeight)/float64(l.maxWeight), nq.value)
+	ag := l.addresses[nq.address]
+	ag.update(nq, flatWeight)
+	l.addressSelect.Update(ag)
+	if valueWeight != 0 {
+		nq.valueWeight = valueWeight
+		l.valueSelect.Update(nq)
+	}
+}
+
+// addToGroup adds the node queue to the given address group. The group is created if
+// it does not exist yet.
+func (l *Limiter) addToGroup(nq *nodeQueue, address string) {
+	nq.address = address
+	ag := l.addresses[address]
+	if ag == nil {
+		ag = &addressGroup{nodeSelect: NewWeightedRandomSelect(flatWeight)}
+		l.addresses[address] = ag
+	}
+	ag.add(nq)
+	l.addressSelect.Update(ag)
+}
+
+// removeFromGroup removes the node queue from its address group
+func (l *Limiter) removeFromGroup(nq *nodeQueue) {
+	ag := l.addresses[nq.address]
+	ag.remove(nq)
+	if len(ag.nodes) == 0 {
+		delete(l.addresses, nq.address)
+	}
+	l.addressSelect.Update(ag)
+}
+
+// remove removes the node queue from its address group, the nodes map and the value
+// selector
+func (l *Limiter) remove(nq *nodeQueue) {
+	l.removeFromGroup(nq)
+	if nq.valueWeight != 0 {
+		l.valueSelect.Remove(nq)
+	}
+	delete(l.nodes, nq.id)
+}
+
+// choose selects the next node queue to process.
+// Note: when a node queue becomes empty it stays in the random selectors for one more
+// selection round before removed, with a weight based on the last relative cost.
+// If no more requests are added to the queue before it is selected again then the
+// queue is removed and the next time a request comes from the same node the queue is
+// added with the highest possible weight.
+func (l *Limiter) choose() *nodeQueue {
+	if l.valueSelect.IsEmpty() || l.selectedByValue {
+		if ag, ok := l.addressSelect.Choose().(*addressGroup); ok {
+			l.selectedByValue = false
+			return ag.choose()
+		}
+	}
+	nq, _ := l.valueSelect.Choose().(*nodeQueue)
```

Comment:
If the valueSelect is empty, should we still set `l.selectedByValue` as true?

Reply:
That flag should be true but the name is indeed misleading in this case. Since it is really about the next selection, not the previous one, I renamed it `selectAddressNext`. Now it makes sense in this case too :)